### ISSUE TITLE
`pytest not installed` when checking h5py

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#156]
I started to install the prereqs for riboviz, and stopped at a point where there appeared to be an issue.

Not a big issue, but for CutAdapt, a different version installed:

 ```console
$ conda install -y -c bioconda cutadapt

Collecting package metadata (current_repodata.json): done

Solving environment: done

## Package Plan ##
  environment location: /home/msachs/miniconda3

  added / updated specs:

    - cutadapt

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bz2file-0.98               |           py37_1          13 KB
    cutadapt-1.18              |   py37h14c3975_1         206 KB  bioconda
    pigz-2.4                   |       h84994c4_0          64 KB
    xopen-0.7.3                |             py_0          11 KB  bioconda
    ------------------------------------------------------------
                                           Total:         294 KB

The following NEW packages will be INSTALLED:

  bz2file            pkgs/main/linux-64::bz2file-0.98-py37_1
  cutadapt           bioconda/linux-64::cutadapt-1.18-py37h14c3975_1
  pigz               pkgs/main/linux-64::pigz-2.4-h84994c4_0
  xopen              bioconda/noarch::xopen-0.7.3-py_0

Downloading and Extracting Packages

pigz-2.4             | 64 KB     | ######################################################### | 100%
cutadapt-1.18        | 206 KB    | ######################################################### | 100%
xopen-0.7.3          | 11 KB     | ######################################################### | 100%
bz2file-0.98         | 13 KB     | ######################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
$ cutadapt --v
1.18
```
I installed Samtools with the explicit command: it also installed an earlier version:

```console
$ conda install -c bioconda samtools

Collecting package metadata (current_repodata.json): done

Solving environment: done

## Package Plan ##

  environment location: /home/msachs/miniconda3
  added / updated specs:
    - samtools

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libgcc-7.2.0               |       h69d50b8_2         269 KB
    samtools-1.7               |                1         1.0 MB  bioconda
    ------------------------------------------------------------
                                           Total:         1.3 MB
The following NEW packages will be INSTALLED:
  libgcc             pkgs/main/linux-64::libgcc-7.2.0-h69d50b8_2
  samtools           bioconda/linux-64::samtools-1.7-1
Proceed ([y]/n)? y
Downloading and Extracting Packages
libgcc-7.2.0         | 269 KB    | ######################################################### | 100%
samtools-1.7         | 1.0 MB    | ######################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
$ samtools --version
samtools 1.7
Using htslib 1.7
Copyright (C) 2018 Genome Research Ltd.
```

More worrisome was that while the subsequent BioPython, gfutils and h5py installations seemed to go without error, there is an error message saying pytest was not installed when I check as per the manual:

```console
$ python
Python 3.7.6 (default, Jan  8 2020, 19:59:22)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> h5py.run_tests()
Tests require pytest, pytest not installed
1
>>> 
```
 
I don’t think that was the expected response to the test…I don’t want to take up your time with these details – but want to be on the same planet as you if you are using riboviz – is there someone I should talk to about these kinds of minutae/errors, which likely will crop up during the installation?

I stopped here with the install just in case there are some dependencies down the line…

**Comment by mikej888** [comment]
I created a copy of my RiboViz environment:
```console
$ conda create --name riboviz-issue --clone base
$ conda activate riboviz-issue
```
Then removed pytest:
```console
$ conda remove -y pytest
```
This now allows the error above to be replicated:
```console
$ python
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import h5py
>>> h5py.run_tests()
Tests require pytest, pytest not installed
1
>>> exit()
```
Installing pytest solves this issue:
```
$ conda install -y pytest
$ python
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
>>> import h5py
>>> h5py.run_tests()
============================= test session starts ==============================
...
tests/test_attribute_create.py ....                                      [  0%]
tests/test_attrs.py ................s                                    [  3%]
tests/test_attrs_data.py ....................                            [  7%]
tests/test_base.py .....                                                 [  8%]
tests/test_completions.py ..                                             [  8%]
tests/test_dataset.py .........x...............................s........ [ 17%]
.........................x..........................                     [ 27%]
tests/test_dataset_getitem.py .......................................... [ 34%]
.......x....................................                             [ 42%]
tests/test_dataset_swmr.py ssss..........                                [ 45%]
...
============ 521 passed, 25 skipped, 3 xfailed, 1 warning in 5.36s =============
0
>>> exit()
```
Note that it states that 3 tests fail. These tests also fail in my standard RiboViz environment that I've been using for a while. The same 2 test modules have test failures:

* Visit https://github.com/h5py/h5py
* Click first "build passing" image on README.md to be taken to TravisCI continuous integration server, https://travis-ci.org/github/h5py/h5py
  - For the last job "master Merge pull request #1520 from weninc/master..." click "2015-passed" which leads to https://travis-ci.org/github/h5py/h5py/builds/666296575
  - Click 2015.1 Python 3.7 which leads to https://travis-ci.org/github/h5py/h5py/jobs/666296576
  - Scroll down output and 4 test failures are noted

```
py37-test-deps/lib/python3.7/site-packages/h5py/tests/test_dataset.py .. [  8%]
......x..x...................................s.......................... [ 21%]
..............x.....................................                     [ 30%]
py37-test-deps/lib/python3.7/site-packages/h5py/tests/test_dataset_getitem.py . [ 30%]
................................................x....................... [ 42%]
..............                                                           [ 45%]
```

* Click second "build passing" image on README.md to be taken to Appveyor continuous integration server, https://ci.appveyor.com/project/h5py/h5py/branch/master
  - Click first job in table to be taken to https://ci.appveyor.com/project/h5py/h5py/branch/master/job/sjg55oyb7vm7x0d0
  - Scroll down output and 3 test failures are noted, again in the same 2 modules:

```
py36-test-deps\lib\site-packages\h5py\tests\test_dataset.py .s.........x [ 10%]
...........................ssssss..s.................................... [ 23%]
....x...................................s.                               [ 30%]
py36-test-deps\lib\site-packages\h5py\tests\test_dataset_getitem.py .... [ 30%]
.............................................x.......................... [ 43%]
..........s                                                              [ 45%]
```

I went back 5 months and picked another job on AppVeyor, https://ci.appveyor.com/project/h5py/h5py/builds/28264247/job/dnkbn8fwiqiap5q1, and the same 2 test modules as above had 3 test failures.

```
py36-test-deps\lib\site-packages\h5py\tests\test_dataset.py .s.........x [ 10%]
...........................ssssss..s.................................... [ 23%]
.x...........................s.                                          [ 29%]
py36-test-deps\lib\site-packages\h5py\tests\test_dataset_getitem.py .... [ 29%]
.............................................x.......................... [ 42%]
.........s                                                               [ 44%]
```

@ewallace, the current user can run

```console
$ conda install -y pytest
```

then continue with RiboViz.

For RiboViz users, the documentation should be updated to:

* State that `pytest` needs to be installed before running `h5py.run_tests()`.
* State that any test failures can be ignored so long as the exit code from the tests is 0.

I'll raise an issue with h5py that updating their [Installation](http://docs.h5py.org/en/stable/build.html#install) docs with info on expected failures would be very useful (we don't really want to have to document third-party package issues like this in the RiboViz docs).

Arguably I'd suggest dropping the instruction to our users to run `h5py.run_tests()`  (as we do comment that "Only minimal installation instructions are given for each prerequisite. See the documentation for each prerequisite for full instructions.") Instead we could recommend that users clone the latest "regression-test-data" repository and run the vignette regression test.

**Comment by mikej888** [comment]
> I'll raise an issue with h5py that updating their Installation docs with info on expected failures would be very useful (we don't really want to have to document third-party package issues like this in the RiboViz docs).

I checked their issue tracker to see about information on expected failures. In passing I found [Test failure on i686 #1337](https://github.com/h5py/h5py/issues/1337] and noticed that the test failure for that user had an exception trace. Relooking at the test reports above I see it is `3 xfailed` and not `3 failed`. pytest [Skip and xfail: dealing with tests that cannot succeed](http://doc.pytest.org/en/latest/skipping.html) explains that this is an "expected failure" (I didn't know pytest supported that concept)
